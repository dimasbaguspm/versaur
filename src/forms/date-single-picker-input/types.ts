/**
 * Props for DateSinglePickerInput
 */

import type { TextInputProps } from '../text-input/types'

/**
 * Props for DateSinglePickerInput
 * Extends TextInputProps, but restricts type to 'date' and value to string
 */
export interface DateSinglePickerInputProps
  extends Omit<TextInputProps, 'type' | 'value' | 'onChange'> {
  /**
   * The value of the input (ISO date string: YYYY-MM-DD)
   */
  value: string
  /**
   * Called when the value changes
   */
  onChange: (value: string) => void
  /**
   * Optional custom formatter for displaying the date value
   * @param value ISO date string
   * @returns formatted string for display
   */
  formatter?: (value: string) => string
}
