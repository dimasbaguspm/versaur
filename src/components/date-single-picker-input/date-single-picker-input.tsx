import React from 'react'

import type { DateSinglePickerInputProps } from './types'
import { DateSinglePickerTrigger } from './date-single-picker-input.atoms'
import { DateSinglePickerDocked } from './date-single-picker-input.docked'
import { DateSinglePickerModal } from './date-single-picker-input.modal'
import { useDateSinglePicker } from './use-date-single-picker'

/**
 * DateSinglePickerInput component for Versaur UI
 *
 * Styled like TextInput, but acts as a button to pick a single date
 * Clicking opens a docked Calendar below the input
 * Strictly typed, accessible, and visually consistent with TextInput
 */
export const DateSinglePickerInput = React.forwardRef<
  HTMLButtonElement,
  DateSinglePickerInputProps
>(
  (
    {
      value,
      onChange,
      label,
      leftContent,
      rightContent,
      helperText,
      error,
      variant = 'primary',
      disabled,
      placeholder = 'Select date',
      id,
      type = 'docked',
      ...props
    },
    ref
  ) => {
    // Controlled: value and onChange are required
    if (value === undefined || typeof onChange !== 'function') {
      throw new Error(
        'DateSinglePickerInput is a controlled component: value and onChange are required.'
      )
    }
    const {
      open,
      setOpen,
      inputId,
      buttonRef,
      handleKeyDown,
      handleMenuClose,
    } = useDateSinglePicker({ id })

    const hasError = Boolean(error)

    // Use consumer's formatDate or fallback
    const defaultFormatDate = (date?: Date) =>
      date
        ? date.toLocaleDateString(undefined, {
            year: 'numeric',
            month: 'short',
            day: 'numeric',
          })
        : typeof placeholder === 'string'
          ? placeholder
          : 'Select date'

    const getFormattedDate = (date?: Date) => {
      if (typeof props.formatDate === 'function') {
        return props.formatDate(date)
      }
      return defaultFormatDate(date)
    }

    // Event emitters
    const handleTriggerClick = () => setOpen(v => !v)
    const handleTriggerKeyDown = handleKeyDown
    const handleDateChange = (date: Date) => {
      onChange?.(date)
      buttonRef.current?.focus()
    }

    return (
      <div className='relative w-full'>
        <div className='relative'>
          {label && (
            <label
              htmlFor={inputId}
              className='block text-sm font-medium text-foreground mb-2'
            >
              {label}
            </label>
          )}
          <DateSinglePickerTrigger
            ref={el => {
              buttonRef.current = el
              if (typeof ref === 'function') ref(el)
              else if (ref) {
                ;(
                  ref as React.MutableRefObject<HTMLButtonElement | null>
                ).current = el
              }
            }}
            id={inputId}
            aria-haspopup='dialog'
            aria-expanded={open}
            aria-controls={open ? `${inputId}-calendar` : undefined}
            aria-invalid={hasError}
            aria-disabled={disabled}
            disabled={disabled}
            leftContent={leftContent}
            rightContent={rightContent}
            hasError={hasError}
            variant={variant}
            value={value}
            placeholder={
              typeof placeholder === 'string' ? placeholder : 'Select date'
            }
            formatDate={getFormattedDate}
            onClick={handleTriggerClick}
            onKeyDown={handleTriggerKeyDown}
            tabIndex={0}
            {...props}
          />
          {hasError && (
            <div className='mt-1 text-sm text-danger' role='alert'>
              {error}
            </div>
          )}
          {helperText && (
            <div className='mt-1 text-sm text-muted-foreground'>
              {helperText}
            </div>
          )}
        </div>

        {type === 'docked' ? (
          <DateSinglePickerDocked
            open={open}
            value={value}
            onChange={handleDateChange}
            inputId={inputId}
            handleMenuClose={handleMenuClose}
          />
        ) : (
          <DateSinglePickerModal
            open={open}
            setOpen={setOpen}
            value={value}
            onChange={handleDateChange}
            label={typeof label === 'string' ? label : undefined}
            buttonRef={buttonRef as React.RefObject<HTMLButtonElement>}
          />
        )}
      </div>
    )
  }
)
