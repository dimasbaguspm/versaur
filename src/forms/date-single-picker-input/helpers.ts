/**
 * Default date formatter for DateSinglePickerInput
 * Formats ISO date string (YYYY-MM-DD) to locale string (e.g., Jul 29, 2025)
 */
export function defaultDateFormatter(date: string): string {
  if (!date) return ''
  const d = new Date(date)
  if (isNaN(d.getTime())) return date
  return d.toLocaleDateString(undefined, {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  })
}
