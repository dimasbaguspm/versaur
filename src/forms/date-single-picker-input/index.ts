export { DateSinglePickerInput } from './date-single-picker-input'
export * from './types'
