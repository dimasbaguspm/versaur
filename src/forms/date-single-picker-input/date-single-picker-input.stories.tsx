/**
 * Stories for DateSinglePickerInput
 * Group: Forms
 * Shows usage and variations for date picking
 */
import { useState } from 'react'
import { DateSinglePickerInput } from './date-single-picker-input'
import type { Meta, StoryObj } from '@storybook/react'

const meta: Meta<typeof DateSinglePickerInput> = {
  title: 'Forms/DateSinglePickerInput',
  component: DateSinglePickerInput,
  parameters: {
    docs: {
      description: {
        component:
          'A wrapper around TextInput for picking a single date. Uses native <input type="date"> for accessibility and browser support.',
      },
    },
  },
}
export default meta

type Story = StoryObj<typeof DateSinglePickerInput>

export const Default: Story = {
  render: function Default() {
    const [date, setDate] = useState('2025-07-29')
    return (
      <DateSinglePickerInput
        value={date}
        onChange={setDate}
        label='Date of Birth'
        helperText='Select your birth date.'
      />
    )
  },
}

export const CustomFormatter: Story = {
  render: function CustomFormatter() {
    const [date, setDate] = useState('2025-08-01')
    return (
      <DateSinglePickerInput
        value={date}
        onChange={setDate}
        label='Date of Birth'
        formatter={date => {
          if (!date) return ''
          const d = new Date(date)
          return d.toLocaleDateString('en-US', {
            month: 'short',
            day: 'numeric',
            year: 'numeric',
          })
        }}
      />
    )
  },
}

export const WithError: Story = {
  render: function WithError() {
    const [date, setDate] = useState('')
    return (
      <DateSinglePickerInput
        value={date}
        onChange={setDate}
        label='Date of Birth'
        error={date === '' ? 'Date is required' : undefined}
        required
      />
    )
  },
}
