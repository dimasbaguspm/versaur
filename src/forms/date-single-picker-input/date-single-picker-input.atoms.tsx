/**
 * Atoms for DateSinglePickerInput
 * - DateSinglePickerTrigger: styled button for input
 * - DateSinglePickerDock: docked calendar container
 */

import { Modal } from '@/overlays/modal'
import { Button } from '@/primitive/button'
import type { DateSinglePickerModalFooterProps } from './types'

import React from 'react'
import { cn } from '@/utils/cn'
import { dateSinglePickerInputVariants } from './helpers'
import type { DateSinglePickerInputProps } from './types'

export interface DateSinglePickerTriggerProps
  extends Omit<React.ButtonHTMLAttributes<HTMLButtonElement>, 'value'> {
  leftContent?: React.ReactNode
  rightContent?: React.ReactNode
  hasError?: boolean
  variant?: DateSinglePickerInputProps['variant']
  value?: Date
  placeholder?: string
  formatDate?: (date?: Date) => string
}

export const DateSinglePickerTrigger = React.forwardRef<
  HTMLButtonElement,
  DateSinglePickerTriggerProps
>(
  (
    {
      leftContent,
      rightContent,
      hasError,
      variant = 'primary',
      value,
      placeholder = 'Select date',
      formatDate,
      className,
      ...props
    },
    ref
  ) => (
    <button
      ref={ref}
      type='button'
      aria-invalid={hasError}
      className={cn(
        dateSinglePickerInputVariants({
          variant: hasError ? 'danger' : variant,
        }),
        leftContent ? 'pl-9' : 'pl-3',
        rightContent ? 'pr-9' : 'pr-3',
        'h-10 flex items-center justify-between cursor-pointer w-full relative',
        'text-left',
        className
      )}
      {...props}
    >
      {leftContent && (
        <span
          className='absolute left-2.5 top-0 bottom-0 pointer-events-none text-gray-500 flex items-center justify-center w-5'
          data-testid='left-content'
        >
          {leftContent}
        </span>
      )}
      <span
        className={cn(
          'flex-1',
          leftContent ? 'ml-2' : '',
          rightContent ? 'mr-2' : ''
        )}
      >
        {formatDate ? formatDate(value) : placeholder}
      </span>
      {rightContent && (
        <span
          className='absolute right-2.5 top-0 bottom-0 pointer-events-none text-gray-500 flex items-center justify-center w-5'
          data-testid='right-content'
        >
          {rightContent}
        </span>
      )}
    </button>
  )
)
DateSinglePickerTrigger.displayName = 'DateSinglePickerTrigger'

export const DateSinglePickerModalFooter = ({
  onCancel,
  onConfirm,
  confirmDisabled,
}: DateSinglePickerModalFooterProps) => (
  <Modal.Footer>
    <Button variant='ghost' onClick={onCancel} type='button'>
      Cancel
    </Button>
    <Button
      variant='primary'
      onClick={onConfirm}
      type='button'
      disabled={confirmDisabled}
    >
      Confirm
    </Button>
  </Modal.Footer>
)
