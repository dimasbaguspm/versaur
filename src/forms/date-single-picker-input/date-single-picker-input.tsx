/**
 * DateSinglePickerInput renders a visually accessible text input and a hidden date input
 * Clicking the visible input triggers the browser date picker
 * The forwarded ref is attached to the hidden date input
 */
import React, { useRef } from 'react'
import type { DateSinglePickerInputProps } from './types'
import { TextInput } from '../text-input'
import { Icon } from '@/primitive/icon'
import { Calendar } from 'lucide-react'
import { defaultDateFormatter } from './helpers'

export const DateSinglePickerInput = React.forwardRef<
  HTMLInputElement,
  DateSinglePickerInputProps
>(function DateSinglePickerInput(
  { value = '', onChange, label, formatter, ...rest },
  ref
) {
  const dateInputRef = useRef<HTMLInputElement>(null)

  // Combine forwardedRef and dateInputRef
  const setRef = (el: HTMLInputElement | null) => {
    dateInputRef.current = el
    if (typeof ref === 'function') {
      ref(el)
    } else if (ref && typeof ref === 'object') {
      ;(ref as React.MutableRefObject<HTMLInputElement | null>).current = el
    }
  }

  // Show the browser date picker when the visible input is clicked
  const handleTextClick = () => {
    if (dateInputRef.current) {
      if (typeof dateInputRef.current.showPicker === 'function') {
        dateInputRef.current.showPicker()
      } else {
        dateInputRef.current.focus()
      }
    }
  }

  // Sync value changes from the hidden date input
  const handleDateChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    onChange(e.target.value)
  }

  const displayValue = (formatter ?? defaultDateFormatter)(value)

  return (
    <>
      <TextInput
        type='text'
        value={displayValue}
        label={label}
        readOnly
        tabIndex={0}
        aria-hidden='true'
        onClick={handleTextClick}
        leftContent={<Icon as={Calendar} color='inherit' size='sm' />}
        data-testid='date-single-picker-visible-input'
        {...rest}
      />
      <input
        ref={setRef}
        type='date'
        value={value}
        onChange={handleDateChange}
        style={{
          position: 'absolute',
          opacity: 0,
          pointerEvents: 'none',
          width: 0,
          height: 0,
        }}
        tabIndex={-1}
        aria-label={typeof label === 'string' ? label : undefined}
      />
    </>
  )
})
