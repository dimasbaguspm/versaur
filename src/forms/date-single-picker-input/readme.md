# DateSinglePickerInput

A simple wrapper around `TextInput` for selecting a single date using native browser `<input type="date">`.

## Usage
```tsx
import { DateSinglePickerInput } from '@dimasbaguspm/versaur/forms'

<DateSinglePickerInput value={date} onChange={setDate} label="Pick a date" />
```

## Props
See `types.ts` for all available props.
