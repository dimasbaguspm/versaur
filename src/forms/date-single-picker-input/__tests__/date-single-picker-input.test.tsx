import { render, screen, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../date-single-picker-input.stories'

describe('DateSinglePickerInput', () => {
  const { Default, WithError, CustomFormatter } = composeStories(stories)

  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(<Default />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('shows label and helper text', () => {
    render(<Default />)
    // There should be two inputs: visible and hidden
    const dateInputs = screen.getAllByLabelText('Date of Birth')
    expect(dateInputs.length).toBeGreaterThanOrEqual(1)
    expect(screen.getByText('Select your birth date.')).toBeInTheDocument()
  })

  it('calls onChange with correct value', () => {
    render(<Default />)
    // The accessible input is type="date" and should be detected by label
    const dateInputs = screen.getAllByLabelText(
      'Date of Birth'
    ) as HTMLInputElement[]
    // The hidden input is type="date"
    const input = dateInputs.find(i => i.type === 'date')!
    fireEvent.change(input, { target: { value: '2025-08-01' } })
    expect(input.value).toBe('2025-08-01')
    // The visible input should show the formatted value
    const visibleInput = screen.getByTestId(
      'date-single-picker-visible-input'
    ) as HTMLInputElement
    expect(visibleInput.value).toBe('Aug 1, 2025')
  })

  it('shows error when value is empty and required', () => {
    render(<WithError />)
    expect(screen.getByText('Date is required')).toBeInTheDocument()
  })

  it('supports custom formatter', () => {
    render(<CustomFormatter />)
    // The visible input should use the custom formatted value
    const visibleInput = screen.getByTestId(
      'date-single-picker-visible-input'
    ) as HTMLInputElement
    expect(visibleInput.value).toBe('Aug 1, 2025')
  })

  it('visible input is readonly and not accessible', () => {
    render(<Default />)
    const visibleInput = screen.getByTestId(
      'date-single-picker-visible-input'
    ) as HTMLInputElement
    expect(visibleInput).toHaveAttribute('readonly')
    expect(visibleInput).not.toHaveAttribute('aria-label')
    expect(visibleInput).not.toHaveAttribute('aria-hidden', 'false')
    expect(visibleInput).toHaveAttribute('aria-hidden', 'true')
  })

  it('hidden input is accessible and receives focus', () => {
    render(<Default />)
    const dateInputs = screen.getAllByLabelText(
      'Date of Birth'
    ) as HTMLInputElement[]
    const input = dateInputs.find(i => i.type === 'date')!
    input.focus()
    expect(document.activeElement).toBe(input)
  })
})
