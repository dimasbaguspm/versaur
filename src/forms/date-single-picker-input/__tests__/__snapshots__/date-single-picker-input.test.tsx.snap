// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateSinglePickerInput > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="w-full"
    >
      <label
        class="block text-sm font-medium text-foreground mb-2"
        for="«r0»"
      >
        Date of Birth
      </label>
      <div
        class="relative w-full"
      >
        <span
          aria-hidden="true"
          class="absolute left-2.5 top-0 bottom-0 pointer-events-none text-gray-500 flex items-center justify-center w-5"
          data-testid="left-content"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-calendar inline-flex items-center justify-center w-4 h-4"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 2v4"
            />
            <path
              d="M16 2v4"
            />
            <rect
              height="18"
              rx="2"
              width="18"
              x="3"
              y="4"
            />
            <path
              d="M3 10h18"
            />
          </svg>
        </span>
        <input
          aria-describedby="«r0»-helper"
          aria-hidden="true"
          aria-invalid="false"
          class="block w-full rounded-md border transition-colors focus:outline-none focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none disabled:bg-gray-50 border-gray-300 bg-gray-50 text-foreground cursor-default focus:ring-0 focus:border-gray-300 pl-9 pr-3 h-9"
          data-testid="date-single-picker-visible-input"
          id="«r0»"
          readonly=""
          tabindex="0"
          type="text"
          value="Jul 29, 2025"
        />
      </div>
      <div
        class="mt-1 text-sm text-gray-600"
        id="«r0»-helper"
      >
        Select your birth date.
      </div>
    </div>
    <input
      aria-label="Date of Birth"
      class="sr-only absolute opacity-0 h-0 w-0 pointer-events-none"
      tabindex="-1"
      type="date"
      value="2025-07-29"
    />
  </div>
</DocumentFragment>
`;
